### ISSUE TITLE
Put icons and labels in single style object to allow decluttering

### ISSUE THREAD

**Issue by mike-000** [openlayers/ol-mapbox-style#279]
I've been testing the newly released OS Vector Tile API https://osdatahub.os.uk/docs/vts/gettingStarted with olms and it needs declutter set on the layer to produce good results.  The downside of that is, for example, rail station icons and labels declutter each other.

**Comment by ahocevar** [comment]
@mike-000 Thanks for reporting this. Icon and label should be on the same style already, if the icon and label style are for the same feature. If not, then there is a bug somewhere. Can you create a test case that fails?

**Comment by mike-000** [comment]
This is based on the OS example code with build versions updated and an info box added for debugging https://codepen.io/mike-000/pen/RwrryMv Change `declutter: false` to true and the station icon at the center of the view disappears leaving only the label.

The olms version is 6.1.1 as I've just noticed 6.1.2 legacy build appears to have been broken by #276

**Comment by ahocevar** [comment]
That codepen does not help really. We'd need a minimal example that shows the issue. Mostly to verify if railway icon and the label are styles of the same feature.

**Comment by mike-000** [comment]
I've updated the info display to show all features at the pixel, plus the available render feature geometry info.  Even if two features had identical geometry and properties one should be displayed with both icon and label when decluttered.

**Comment by ahocevar** [comment]
> Even if two features had identical geometry and properties one should be displayed with both icon and label when decluttered.

That would be very expensive during style function execution.

First, I'd like to see what is going on here. Please share the relevant sections of the style, i.e. the style that is applied to the features in question. Note that both the icon and the text also have to be in the same style (`icon-image` and `text-field` in the same `layout` section.

**Comment by mike-000** [comment]
I have added logging of the feature and the style function output and a single feature is getting a style array.
```

{
  "type": "Point",
  "extent": [
    628715,
    250368.125,
    628715,
    250368.125
  ],
  "properties": {
    "_symbol": 4,
    "_label_class": 0,
    "_name": "Melton",
    "_label_class1": 1,
    "_name1": "Melton",
    "layer": "Railway_stations"
  }
} [
  {
    "fill_": null,
    "image_": {
      "opacity_": 1,
      "rotateWithView_": false,
      "rotation_": 0,
      "scale_": 0.9,
      "displacement_": [
        0,
        0
      ],
      "anchor_": [
        0.5,
        0.5
      ],
      "normalizedAnchor_": null,
      "anchorOrigin_": "top-left",
      "anchorXUnits_": "fraction",
      "anchorYUnits_": "fraction",
      "crossOrigin_": null,
      "color_": null,
      "iconImage_": {
        "disposed_": false,
        "pendingRemovals_": {},
        "dispatching_": {},
        "listeners_": {},
        "hitDetectionImage_": {},
        "image_": {},
        "canvas_": null,
        "color_": null,
        "unlisten_": null,
        "imageState_": 2,
        "size_": [
          335,
          286
        ],
        "src_": "https://api.os.uk/maps/vector/v1/vts/resources/sprites/sprite.png?key=ufArYa1UUPHJcOYbiJDaKkA7Fb4oCkEs",
        "tainted_": false
      },
      "offset_": [
        282,
        201
      ],
      "offsetOrigin_": "top-left",
      "origin_": [
        282,
        201
      ],
      "size_": [
        26,
        18
      ]
    },
    "renderer_": null,
    "stroke_": null,
    "zIndex_": 194
  },
  {
    "fill_": null,
    "renderer_": null,
    "stroke_": null,
    "text_": {
      "font_": "normal 400 13px/1.2 \"Source Sans Pro\"",
      "rotation_": 0,
      "rotateWithView_": false,
      "text_": "M e l t o n",
      "textAlign_": "center",
      "textBaseline_": "top",
      "fill_": {
        "color_": "rgba(0,0,0,1)"
      },
      "maxAngle_": 0.7853981633974483,
      "placement_": "point",
      "overflow_": false,
      "stroke_": {
        "color_": "rgba(250,250,243,1)",
        "lineDash_": null,
        "width_": 3
      },
      "offsetX_": 0,
      "offsetY_": 11.031794999999999,
      "backgroundFill_": null,
      "backgroundStroke_": null,
      "padding_": [
        2,
        2,
        2,
        2
      ]
    },
    "zIndex_": 339
  }
]

```

The style is defined by https://api.os.uk/maps/vector/v1/vts/resources/styles?key=ufArYa1UUPHJcOYbiJDaKkA7Fb4oCkEs


```
This is the symbol definition:

    {
      "id": "OS/Railway_stations/Railway Station",
      "type": "symbol",
      "source": "esri",
      "source-layer": "Railway_stations",
      "filter": [
        "==",
        "_symbol",
        4
      ],
      "minzoom": 7,
      "maxzoom": 12,
      "layout": {
        "icon-image": "OS/Railway_stations/Railway Station",
        "icon-size": {
          "stops": [
            [
              8,
              0.6
            ],
            [
              9,
              0.7
            ],
            [
              10,
              0.8
            ],
            [
              11,
              0.9
            ],
            [
              12,
              0.9
            ]
          ]
        },
        "icon-allow-overlap": true
      },
      "paint": {}
    },

```
and these are the labels

```

    {
      "id": "OS/Railway_stations/label/Station names 22400-89600",
      "type": "symbol",
      "source": "esri",
      "source-layer": "Railway_stations",
      "filter": [
        "==",
        "_label_class1",
        1
      ],
      "minzoom": 7,
      "maxzoom": 9,
      "layout": {
        "text-font": [
          "Source Sans Pro Regular"
        ],
        "text-size": 12,
        "text-anchor": "top",
        "text-offset": {
          "property": "_symbol",
          "type": "categorical",
          "default": [
            0,
            0.0333333
          ],
          "stops": [
            [
              0,
              [
                0,
                0.631292
              ]
            ],
            [
              1,
              [
                0,
                0.359493
              ]
            ],
            [
              2,
              [
                0,
                0.609832
              ]
            ],
            [
              3,
              [
                0,
                0.629975
              ]
            ],
            [
              4,
              [
                0,
                0.625794
              ]
            ],
            [
              5,
              [
                0,
                0.410354
              ]
            ]
          ]
        },
        "text-letter-spacing": 0.05,
        "text-field": "{_name1}",
        "text-optional": true
      },
      "paint": {
        "text-color": "#000000",
        "text-halo-color": "#FAFAF3",
        "text-halo-width": 1.5
      }
    },
    {
      "id": "OS/Railway_stations/label/Station names 2800-22400",
      "type": "symbol",
      "source": "esri",
      "source-layer": "Railway_stations",
      "filter": [
        "==",
        "_label_class",
        0
      ],
      "minzoom": 9,
      "maxzoom": 12,
      "layout": {
        "text-font": [
          "Source Sans Pro Regular"
        ],
        "text-size": 13,
        "text-anchor": "top",
        "text-offset": {
          "property": "_symbol",
          "type": "categorical",
          "default": [
            0,
            0.1
          ],
          "stops": [
            [
              0,
              [
                0,
                0.638163
              ]
            ],
            [
              1,
              [
                0,
                0.393543
              ]
            ],
            [
              2,
              [
                0,
                0.618849
              ]
            ],
            [
              3,
              [
                0,
                0.636977
              ]
            ],
            [
              4,
              [
                0,
                0.633215
              ]
            ],
            [
              5,
              [
                0,
                0.439319
              ]
            ]
          ]
        },
        "text-letter-spacing": 0.05,
        "text-field": "{_name}",
        "text-optional": true
      },
      "paint": {
        "text-color": "#000000",
        "text-halo-color": "#FAFAF3",
        "text-halo-width": 1.5
      }
    },

```

**Comment by ahocevar** [comment]
Thanks for the thorough investigation, @mike-000.

So icon and label are on different styles. That explains the behaviour you are observing. I wonder how mapbox-gl-js handles such a case. If they handle this case as you expect, we might consider changing that in ol-mapbox-style to behave the same. It would be doable without added performance penalty.

**Comment by mike-000** [comment]
I cannot think of any instances where a label and an icon for the same feature should declutter one another.

There is a Mapbox example at https://labs.os.uk/public/os-data-hub-examples/os-vector-tile-api/vts-3857-basic-map, similar data but different projection.  Icons are not decluttered, but the label may be decluttered by a more significant one which might be rendered over the icon.

![image](https://user-images.githubusercontent.com/49240900/86623096-e24ff500-bfb8-11ea-8a30-ff52d6d661f2.png)

I have also noticed Mapbox behaves differently when overzooming.  If the max tile zoom is 11 (i.e. no access to premium data) and the style says maxzoom 12 features disappear when overzoomed, which does not happen with olms,  I actually prefer the olms behaviour and have added code to loop through the style object layers deleting maxzoom entries above 11 when using Mapbox.

**Comment by ahocevar** [comment]
I took a closer look now at the styles you pasted. The symbol style has `"icon-allow-overlap": true` set, which means the icon won't be decluttered. This is something that ol-mapbox-style cannot do easily, because decluttering in OpenLayers is layer based and not style based. This would have to be changed in OpenLayers.

**Comment by mike-000** [comment]
Both icon and label being decluttered by a more significant feature would be an acceptable limitation of OpenLayers.  A style where the label always declutters its own feature icon isn't.  As a workaround I have fixed it in the application but it would be better if it could be done within olms

                olms.stylefunction(layer, styleJson, 'esri', resolutions, spritesJson, spritesPngUrl);
                var styleFunction = layer.getStyleFunction();
                layer.setStyle(function(feature, resolution) {
                  var style = styleFunction(feature, resolution);
                  if (!style || style.length !== 2) {
                    return style;
                  }
                  var zIndex = Math.max(style[0].getZIndex(), style[1].getZIndex());
                  var text0 = style[0].getText();
                  var text1 = style[1].getText();
                  var image0 = style[0].getImage();
                  var image1 = style[1].getImage();
                  if (text0 && !text1 && image1 && !image0) {
                    var combo = style[1].clone();
                    combo.setText(text0);
                    combo.setZIndex(zIndex);
                    return [combo];
                  } else if (text1 && !text0 && image0 && !image1) {
                    var combo = style[0].clone();
                    combo.setText(text1);
                    combo.setZIndex(zIndex);
                    return [combo];
                  } else {
                    return style;
                  }
                });

**Comment by ahocevar** [comment]
I'm afraid you'll have to keep this expensive workaround in your application. I still believe this should be fixed in OpenLayers and ol-mapbox-style by adding support for the `icon-allow-overlap` option. On the OpenLayers side, adding support for per-style decluttering would be a welcome improvement, and not too hard to implement.

A cheaper workaround would be to change the input style, where you could easily combine icon and label into a single style `layout`, e.g.
```json
"layout": {
        "icon-image": "OS/Railway_stations/Railway Station",
        "icon-size": {
          "stops": [[8, 0.6], [9, 0.7], [10, 0.8], [11, 0.9], [12, 0.9]]
        },
        "icon-allow-overlap": true,
"text-font": ["Source Sans Pro Regular"],
        "text-size": 13,
        "text-anchor": "top",
        "text-offset": {
          "property": "_symbol",
          "type": "categorical",
          "default": [0, 0.1],
          "stops": [[0, [0, 0.638163]], [1, [0, 0.393543]], [2, [0, 0.618849]], [3, [0, 0.636977]], [4, [0, 0.633215]], [5, [0, 0.439319]]]
        },
        "text-letter-spacing": 0.05,
        "text-field": "{_name}",
        "text-optional": true
      },
```
Note that both Mapbox and Maptiler use styles like this to combine icons and text.

**Comment by mike-000** [comment]
Ordnance Survey did use that approach with their Zoomstack styles (as used by Maptiler).  In this case it would be messy to make changes as there are several categories of station which would each need updating, as well as anything else which uses both icon and text.  Even with the workaround performance is acceptable (with additional checks to limit it to point geometry and to ensure text styles contain text), much better than without decluttering where at some resolutions it became overloaded.

**Comment by cns-solutions-admin** [comment]
See also https://github.com/openlayers/openlayers/issues/7475 and https://github.com/openlayers/ol-mapbox-style/pull/561

**Comment by mike-000** [comment]
Fixed by #561

**Comment by mike-000** [comment]
Fixed by #561
